arroyo/types.py

```diff
@@ -32,31 +32,46 @@ class Message(Generic[TPayload]):
     Represents a single message within a partition.
     """
 
-    __slots__ = ["partition", "offset", "payload", "timestamp"]
+    __slots__ = ["payload", "committable"]
 
-    partition: Partition
-    offset: int
     payload: TPayload
-    timestamp: datetime
+    committable: Mapping[Partition, Position]
```

Comment:
What if `committable` was a property or a method. 
So you could have a smarter implementation than duplicating the data between `Message` and `BrokerPayload`.

```
def committable(self) -> Mapping[Partition, Position]:
    if isinstance(self.payload, BrokerPayload):
         return {self.payload.partition: self.payload.position}
    elif isinstance(self.payload, BatchPayload):
         return .... whatever it takes to get the proper mapping ....
    # or you could have the dict initialized when the instance is created so you do not need to create a dictionary every time `committable` is called.
```

A cleaner alternative.
- add `committable` as a method of the current `Message`.
- Have subclaasses of Message. The broker based one just returns a dictionary with one partition and one position. The other one is for batch messages.
This would break a bit the concept of payload we have today but I am not sure it would be a big change.

Reply:
Updated to remove the duplication of offsets. Let me know what you think.